systems/controllers/generate_samples.cc

```diff
@@ -0,0 +1,626 @@
+
+    UpdateContext(n_q, n_v, n_u, plant, context, plant_ad, context_ad,
+                  projected_state);
+    if (check_collision(n_q, n_v, n_u, projected_state, plant, context,
+                        plant_ad, context_ad, contact_geoms, sampling_params,
+                        sampling_c3_options, min_distance_index)) {
+      continue;  // If the projected state is still in collision, exit loop and
+                 // generate a new sample.
+    }
+
+    // Generate a new sample if the projected sample is on the top or bottom
+    // surface of the object. i.e not near the sampling height.
+    double epsilon = 0.001;
+    if (projected_state[2] < -0.01) {
+      continue;
+    }
+
+    // Undo the update context.
+    UpdateContext(n_q, n_v, n_u, plant, context, plant_ad, context_ad, x_lcs);
+    return projected_state;
+  }
+}
+
+bool check_collision(
+    const int& n_q, const int& n_v, const int& n_u,
+    const Eigen::VectorXd& candidate_state,
+    drake::multibody::MultibodyPlant<double>& plant,
+    drake::systems::Context<double>* context,
+    drake::multibody::MultibodyPlant<drake::AutoDiffXd>& plant_ad,
+    drake::systems::Context<drake::AutoDiffXd>* context_ad,
+    const std::vector<
+        std::vector<drake::SortedPair<drake::geometry::GeometryId>>>&
+        contact_geoms,
+    const SamplingC3SamplingParams& sampling_params,
+    SamplingC3Options sampling_c3_options, int& min_distance_index) {
+  // This function returns a boolean value indicating whether the sample is in
+  // collision with the object or not. If the sample is in collision with the
+  // object, the function modifies a reference to the index of the closest pair
+  // of contact geoms so that the projection function need not recompute the
+  // closest pair.
+
+  // Update the context of the plant with the candidate state.
+  UpdateContext(n_q, n_v, n_u, plant, context, plant_ad, context_ad,
+                candidate_state);
+
+  // Find the closest pair if there are multiple pairs
+  std::vector<double> distances;
+
+  for (int i = 0; i < contact_geoms.at(0).size(); i++) {
+    // Evaluate the distance for each pair
+    SortedPair<GeometryId> pair{(contact_geoms.at(0)).at(i)};
+    multibody::GeomGeomCollider collider(plant, pair);
+
+    auto [phi_i, J_i] = collider.EvalPolytope(
+        *context, sampling_c3_options.num_friction_directions);
+    distances.push_back(phi_i);
+  }
+
+  // Find the minimum distance.
+  auto min_distance_it = std::min_element(distances.begin(), distances.end());
+  // This line modifies the min_distance_index variable passed by reference.
+  min_distance_index = std::distance(distances.begin(), min_distance_it);
+  double min_distance = *min_distance_it;
+
+  return min_distance <= sampling_params.sample_projection_clearance - 1e-3;
+}
+
+// Helper function to update context of a plant with a given state.
+void UpdateContext(const int& n_q, const int& n_v, const int& n_u,
```

Comment:
I don't love that this is a duplicate of the function in our sampling controller.  Could we think of a way to make this a more generally accessible function in a standalone file that both things call?  I know the controller's implementation accesses class variables, but those can be passed in like how this one is used.